rcl/test/rcl/test_events.cpp

```diff
@@ -0,0 +1,546 @@
+
+    // init publisher
+    ret = setup_publisher(deadline, lifespan, lease_duration, liveliness_policy);
+    ASSERT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+
+    // init publisher events
+    publisher_event = rcl_get_zero_initialized_event();
+    ret = rcl_publisher_event_init(&publisher_event, &publisher, pub_event_type);
+    ASSERT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+
+    // init subscription
+    ret = setup_subscriber(deadline, lifespan, lease_duration, liveliness_policy);
+    ASSERT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+
+    // init subscription event
+    subscription_event = rcl_get_zero_initialized_event();
+    ret = rcl_subscription_event_init(&subscription_event, &subscription, sub_event_type);
+    ASSERT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+
+    // wait for discovery
+    // total wait time of 10 seconds, if never ready
+    size_t max_iterations = 100;
+    milliseconds wait_period(10);
+    size_t iteration = 0;
+    do {
+      iteration++;
+      size_t count = 0;
+      rcl_ret_t ret = rcl_subscription_get_publisher_count(&subscription, &count);
+      ASSERT_EQ(ret, RCL_RET_OK);
+      if (count > 0) {
+        break;
+      }
+      std::this_thread::sleep_for(wait_period);
+    } while (iteration < max_iterations);
+  }
+
+  void tear_down_publisher_subscriber()
+  {
+    rcl_ret_t ret;
+
+    ret = rcl_event_fini(&subscription_event);
+    EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+
+    ret = rcl_subscription_fini(&subscription, this->node_ptr);
+    EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+
+    ret = rcl_event_fini(&publisher_event);
+    EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+
+    ret = rcl_publisher_fini(&publisher, this->node_ptr);
+    EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+  }
+
+  void TearDown()
+  {
+    rcl_ret_t ret;
+
+    ret = rcl_node_fini(this->node_ptr);
+    delete this->node_ptr;
+    EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+
+    ret = rcl_shutdown(this->context_ptr);
+    EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+
+    ret = rcl_context_fini(this->context_ptr);
+    delete this->context_ptr;
+    EXPECT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
+  }
+
+protected:
+  rcl_context_t * context_ptr;
+  rcl_node_t * node_ptr;
+
+  rcl_publisher_t publisher;
+  rcl_event_t publisher_event;
+  rcl_subscription_t subscription;
+  rcl_event_t subscription_event;
+  bool is_unsupported, is_opensplice, is_fastrtps;
+  const char * topic = "rcl_test_publisher_subscription_events";
+  const rosidl_message_type_support_t * ts;
+};
+
+rcl_ret_t
+wait_for_msgs_and_events(
+  rcl_subscription_t * subscription,
+  rcl_event_t * subscription_event,
+  rcl_event_t * publisher_event,
+  rcl_context_t * context,
+  int64_t period_ms,
+  bool & msg_ready,
```

Comment:
Can you take `bool *` instead here? The current state leads to misleading usage, e.g.:

```c++
  // wait for events
  bool msg_ready, subscription_event_ready, publisher_event_ready;

  // wait for lease duration to expire
  std::this_thread::sleep_for(DEADLINE_PERIOD_IN_S + milliseconds(500));
  ASSERT_EQ(wait_for_msgs_and_events(&subscription, &subscription_event, &publisher_event,
    context_ptr, 1000, msg_ready, subscription_event_ready, publisher_event_ready), RCL_RET_OK);
//                     ^ I assume a copy is made, or a const & is being used, when passing a value like this
```

Also the Google Style guide for C++ (which our style is based on) does not allow non-const references:

https://google.github.io/styleguide/cppguide.html#Reference_Arguments

Reply:
I will change it to a bool* since the google style guide does not allow non-const references.
 
bool & msg_ready implicitly enforces references to NULL, void, invalid objects or functions, and bit fields. Google's Con for this does not seem to outweigh its benefits "References can be confusing, as they have value syntax but pointer semantics." This enforces the intention of this function and reduces the bloating of invalid param checks.

Reply:
Okay, will do.